File: src/applications/arm_demos/src/mtc_pick_place_demo.cpp
/**
 * @file mtc_pick_place_demo.cpp
 * @brief MoveIt Task Constructor Pick and Place Demo (C++ port)
 *
 * Demonstrates using MTC to perform a pick-and-place task with the
 * 6-DOF humanoid arm. Stages:
 * - CurrentState
 * - Pick container (approach, grasp IK, allow collision, attach, lift)
 * - Connect (move to place)
 * - Place container (lower, place pose, detach, forbid collision, retreat)
 * - Return home
 *
 * Author: LDR Robotics Team
 * License: MIT
 */

#include <rclcpp/rclcpp.hpp>
#include <moveit/planning_scene/planning_scene.hpp>
#include <moveit/planning_scene_interface/planning_scene_interface.hpp>
#include <moveit/task_constructor/task.h>
#include <moveit/task_constructor/solvers.h>
#include <moveit/task_constructor/stages.h>

#if __has_include(<tf2_geometry_msgs/tf2_geometry_msgs.hpp>)
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#else
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#endif

#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/vector3_stamped.hpp>
#include <shape_msgs/msg/solid_primitive.hpp>

namespace mtc = moveit::task_constructor;

static const rclcpp::Logger LOGGER = rclcpp::get_logger("mtc_pick_place_demo");

class MTCPickPlaceDemo
{
public:
  MTCPickPlaceDemo(const rclcpp::NodeOptions& options);

  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr getNodeBaseInterface();

  void run();

private:
  void setupPlanningScene();
  mtc::Task createTask();
  mtc::Task task_;
  rclcpp::Node::SharedPtr node_;

  std::string arm_group_name_;
  std::string hand_group_name_;
  std::string eef_name_;
  std::string hand_frame_;
  std::string object_name_;
  std::string surface_name_;
};

MTCPickPlaceDemo::MTCPickPlaceDemo(const rclcpp::NodeOptions& options)
  : node_(std::make_shared<rclcpp::Node>("mtc_pick_place_demo", options))
{
  arm_group_name_ = "arm";
  hand_group_name_ = "hand";
  eef_name_ = "tcp";
  hand_frame_ = "wrist_roll_link";
  object_name_ = "target_object";
  surface_name_ = "table";

  RCLCPP_INFO(LOGGER, "MTC Pick and Place Demo initialized");
}

rclcpp::node_interfaces::NodeBaseInterface::SharedPtr
MTCPickPlaceDemo::getNodeBaseInterface()
{
  return node_->get_node_base_interface();
}

void MTCPickPlaceDemo::setupPlanningScene()
{
  moveit::planning_interface::PlanningSceneInterface psi;

  // Add a table as support surface
  {
    moveit_msgs::msg::CollisionObject table;
    table.id = surface_name_;
    table.header.frame_id = "base_link";
    table.primitives.resize(1);
    table.primitives[0].type = shape_msgs::msg::SolidPrimitive::BOX;
    table.primitives[0].dimensions = {0.6, 0.6, 0.02};
    table.primitive_poses.resize(1);
    table.primitive_poses[0].position.x = 0.3;
    table.primitive_poses[0].position.y = 0.0;
    table.primitive_poses[0].position.z = 0.19;
    table.primitive_poses[0].orientation.w = 1.0;
    table.operation = moveit_msgs::msg::CollisionObject::ADD;
    psi.applyCollisionObject(table);
  }

  // Add target object (cylinder)
  {
    moveit_msgs::msg::CollisionObject object;
    object.id = object_name_;
    object.header.frame_id = "base_link";
    object.primitives.resize(1);
    object.primitives[0].type = shape_msgs::msg::SolidPrimitive::CYLINDER;
    object.primitives[0].dimensions = {0.1, 0.025};  // height, radius
    object.primitive_poses.resize(1);
    object.primitive_poses[0].position.x = 0.3;
    object.primitive_poses[0].position.y = 0.0;
    object.primitive_poses[0].position.z = 0.25;
    object.primitive_poses[0].orientation.w = 1.0;
    object.operation = moveit_msgs::msg::CollisionObject::ADD;
    psi.applyCollisionObject(object);
  }

  RCLCPP_INFO(LOGGER, "Planning scene set up with table and target object");
}

mtc::Task MTCPickPlaceDemo::createTask()
{
  mtc::Task task;
  task.stages()->setName("pick_place_task");
  task.loadRobotModel(node_);

  // Set task properties
  task.setProperty("group", arm_group_name_);
  task.setProperty("eef", hand_group_name_);
  task.setProperty("ik_frame", hand_frame_);

  // Create planners
  auto sampling_planner = std::make_shared<mtc::solvers::PipelinePlanner>(node_);

  auto cartesian_planner = std::make_shared<mtc::solvers::CartesianPath>();
  cartesian_planner->setMaxVelocityScalingFactor(1.0);
  cartesian_planner->setMaxAccelerationScalingFactor(1.0);
  cartesian_planner->setStepSize(0.01);

  auto interpolation_planner = std::make_shared<mtc::solvers::JointInterpolationPlanner>();

  // Stage 1: Current State
  mtc::Stage* current_state_ptr = nullptr;
  {
    auto stage = std::make_unique<mtc::stages::CurrentState>("current state");
    current_state_ptr = stage.get();
    task.add(std::move(stage));
  }

  // ==========================
  // Pick Object Container
  // ==========================
  {
    auto pick_container = std::make_unique<mtc::SerialContainer>("pick object");
    task.properties().exposeTo(pick_container->properties(), {"eef", "group", "ik_frame"});
    pick_container->properties().configureInitFrom(mtc::Stage::PARENT, {"eef", "group", "ik_frame"});

    // 3a: Approach object
    {
      auto stage = std::make_unique<mtc::stages::MoveRelative>("approach object", cartesian_planner);
      stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group"});
      stage->setMinMaxDistance(0.05, 0.15);

      geometry_msgs::msg::Vector3Stamped direction;
      direction.header.frame_id = "base_link";
      direction.vector.z = -1.0;  // Move down
      stage->setDirection(direction);
      pick_container->insert(std::move(stage));
    }

    // 3b: Generate grasp pose
    {
      auto stage = std::make_unique<mtc::stages::GenerateGraspPose>("generate grasp pose");
      stage->properties().configureInitFrom(mtc::Stage::PARENT);
      stage->setPreGraspPose("open");
      stage->setObject(object_name_);
      stage->setAngleDelta(0.2);
      stage->setMonitoredStage(current_state_ptr);

      // Compute IK for grasp
      auto wrapper = std::make_unique<mtc::stages::ComputeIK>("grasp pose IK", std::move(stage));
      wrapper->setMaxIKSolutions(8);
      wrapper->setIKFrame(hand_frame_);
      wrapper->properties().configureInitFrom(mtc::Stage::PARENT, {"eef", "group"});
      wrapper->properties().configureInitFrom(mtc::Stage::INTERFACE, {"target_pose"});
      pick_container->insert(std::move(wrapper));
    }

    // 3c: Allow collision (hand-object)
    {
      auto stage = std::make_unique<mtc::stages::ModifyPlanningScene>("allow collision (hand-object)");
      stage->allowCollisions(
        object_name_,
        task.getRobotModel()
          ->getJointModelGroup(hand_group_name_)
          ->getLinkModelNamesWithCollisionGeometry(),
        true);
      pick_container->insert(std::move(stage));
    }

    // 3d: Attach object
    {
      auto stage = std::make_unique<mtc::stages::ModifyPlanningScene>("attach object");
      stage->attachObject(object_name_, hand_frame_);
      pick_container->insert(std::move(stage));
    }

    // 3e: Lift object
    {
      auto stage = std::make_unique<mtc::stages::MoveRelative>("lift object", cartesian_planner);
      stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group"});
      stage->setMinMaxDistance(0.08, 0.15);

      geometry_msgs::msg::Vector3Stamped direction;
      direction.header.frame_id = "base_link";
      direction.vector.z = 1.0;  // Lift up
      stage->setDirection(direction);
      pick_container->insert(std::move(stage));
    }

    task.add(std::move(pick_container));
  }

  // ==========================
  // Move to Place
  // ==========================
  {
    auto stage = std::make_unique<mtc::stages::Connect>(
      "move to place",
      mtc::stages::Connect::GroupPlannerVector{
        {arm_group_name_, sampling_planner}
      });
    stage->properties().configureInitFrom(mtc::Stage::PARENT);
    task.add(std::move(stage));
  }

  // ==========================
  // Place Object Container
  // ==========================
  {
    auto place_container = std::make_unique<mtc::SerialContainer>("place object");
    task.properties().exposeTo(place_container->properties(), {"eef", "group", "ik_frame"});
    place_container->properties().configureInitFrom(mtc::Stage::PARENT, {"eef", "group", "ik_frame"});

    // 5a: Lower object
    {
      auto stage = std::make_unique<mtc::stages::MoveRelative>("lower object", cartesian_planner);
      stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group"});
      stage->setMinMaxDistance(0.05, 0.15);

      geometry_msgs::msg::Vector3Stamped direction;
      direction.header.frame_id = "base_link";
      direction.vector.z = -1.0;  // Move down
      stage->setDirection(direction);
      place_container->insert(std::move(stage));
    }

    // 5b: Generate place pose
    {
      auto stage = std::make_unique<mtc::stages::GeneratePlacePose>("generate place pose");
      stage->properties().configureInitFrom(mtc::Stage::PARENT, {"ik_frame"});
      stage->setObject(object_name_);

      geometry_msgs::msg::PoseStamped place_pose;
      place_pose.header.frame_id = "base_link";
      place_pose.pose.position.x = 0.3;
      place_pose.pose.position.y = 0.3;
      place_pose.pose.position.z = 0.25;
      place_pose.pose.orientation.w = 1.0;
      stage->setPose(place_pose);
      stage->setMonitoredStage(current_state_ptr);

      auto wrapper = std::make_unique<mtc::stages::ComputeIK>("place pose IK", std::move(stage));
      wrapper->setMaxIKSolutions(8);
      wrapper->setIKFrame(hand_frame_);
      wrapper->properties().configureInitFrom(mtc::Stage::PARENT, {"eef", "group"});
      wrapper->properties().configureInitFrom(mtc::Stage::INTERFACE, {"target_pose"});
      place_container->insert(std::move(wrapper));
    }

    // 5c: Detach object
    {
      auto stage = std::make_unique<mtc::stages::ModifyPlanningScene>("detach object");
      stage->detachObject(object_name_, hand_frame_);
      place_container->insert(std::move(stage));
    }

    // 5d: Forbid collision (hand-object)
    {
      auto stage = std::make_unique<mtc::stages::ModifyPlanningScene>("forbid collision (hand-object)");
      stage->allowCollisions(
        object_name_,
        task.getRobotModel()
          ->getJointModelGroup(hand_group_name_)
          ->getLinkModelNamesWithCollisionGeometry(),
        false);
      place_container->insert(std::move(stage));
    }

    // 5e: Retreat
    {
      auto stage = std::make_unique<mtc::stages::MoveRelative>("retreat", cartesian_planner);
      stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group"});
      stage->setMinMaxDistance(0.05, 0.15);

      geometry_msgs::msg::Vector3Stamped direction;
      direction.header.frame_id = "base_link";
      direction.vector.z = 1.0;  // Retreat up
      stage->setDirection(direction);
      place_container->insert(std::move(stage));
    }

    task.add(std::move(place_container));
  }

  // ==========================
  // Return Home
  // ==========================
  {
    auto stage = std::make_unique<mtc::stages::MoveTo>("return home", sampling_planner);
    stage->setGroup(arm_group_name_);
    stage->setGoal("home");
    task.add(std::move(stage));
  }

  RCLCPP_INFO(LOGGER, "MTC pick-and-place task created with all stages");
  return task;
}

void MTCPickPlaceDemo::run()
{
  // Setup planning scene
  RCLCPP_INFO(LOGGER, "Setting up planning scene...");
  rclcpp::sleep_for(std::chrono::seconds(2));
  setupPlanningScene();

  // Create task
  RCLCPP_INFO(LOGGER, "Creating task...");
  task_ = createTask();

  // Initialize task
  try {
    task_.init();
    RCLCPP_INFO(LOGGER, "Task initialized successfully");
  } catch (mtc::InitStageException& e) {
    RCLCPP_ERROR_STREAM(LOGGER, "Task initialization failed: " << e);
    return;
  }

  // Plan task
  RCLCPP_INFO(LOGGER, "Planning task...");
  if (!task_.plan(5)) {
    RCLCPP_ERROR(LOGGER, "Task planning failed - no solutions found");
    return;
  }

  RCLCPP_INFO(LOGGER, "Planning succeeded! %zu solution(s) found", task_.solutions().size());

  // Publish solution for visualization
  task_.introspection().publishSolution(*task_.solutions().front());

  // Execute the best solution
  RCLCPP_INFO(LOGGER, "Executing task...");
  auto result = task_.execute(*task_.solutions().front());
  if (result.val != moveit_msgs::msg::MoveItErrorCodes::SUCCESS) {
    RCLCPP_ERROR(LOGGER, "Task execution failed with error code: %d", result.val);
    return;
  }

  RCLCPP_INFO(LOGGER, "Pick-and-place execution completed successfully!");
}

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);

  rclcpp::NodeOptions options;
  options.automatically_declare_parameters_from_overrides(true);

  auto demo = std::make_shared<MTCPickPlaceDemo>(options);

  rclcpp::executors::MultiThreadedExecutor executor;
  executor.add_node(demo->getNodeBaseInterface());

  auto spin_thread = std::make_unique<std::thread>([&executor]() {
    executor.spin();
  });

  // Wait for system to be ready
  RCLCPP_INFO(LOGGER, "Waiting for system to be ready...");
  rclcpp::sleep_for(std::chrono::seconds(3));

  demo->run();

  RCLCPP_INFO(LOGGER, "Demo completed. Keeping node alive for visualization. Press Ctrl+C to exit.");
  spin_thread->join();
  rclcpp::shutdown();
  return 0;
}

File: src/applications/arm_demos/scripts/mtc_pick_place_demo.py
#!/usr/bin/env python3
"""
MoveIt Task Constructor Pick and Place Demo

This script demonstrates using MoveIt Task Constructor (MTC) to perform
a pick-and-place task with the 5-DOF humanoid arm.

MTC allows you to define manipulation tasks as a sequence of stages:
- Approach object
- Grasp object
- Lift object
- Move to target
- Place object
- Retreat

Author: LDR Robotics Team
License: MIT
"""

import sys

# Check if MTC is available
try:
    from moveit.task_constructor import core, stages
    from moveit.core.robot_state import RobotState
except ImportError as e:
    print("\n" + "="*60)
    print("ERROR: MoveIt Task Constructor not installed!")
    print("="*60)
    print(f"\nImport error: {e}")
    print("\nTo install MTC, run:")
    print("  sudo apt install ros-humble-moveit-task-constructor-*")
    print("\nFor detailed installation instructions, run:")
    print("  ros2 run arm_demos check_mtc.py")
    print("\n" + "="*60)
    sys.exit(1)

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped, Pose, Vector3Stamped
from shape_msgs.msg import SolidPrimitive
from moveit_msgs.msg import Constraints, CollisionObject


class MTCPickPlaceDemo(Node):
    """Node for demonstrating MoveIt Task Constructor pick and place."""

    def __init__(self):
        super().__init__('mtc_pick_place_demo')

        # MTC task
        self.task = None

        # Configuration
        self.arm_group_name = "arm"
        self.hand_group_name = "hand"  # Will be used when gripper is added
        self.eef_name = "tcp"  # End-effector name (update when gripper added)
        self.hand_frame = "wrist_roll_link"  # Current end-effector link

        # Object parameters
        self.object_name = "target_object"
        self.surface_name = "table"

        self.get_logger().info("MTC Pick and Place Demo initialized")

    def setup_planning_scene(self):
        """Add collision objects to the planning scene."""
        # This would typically use PlanningSceneInterface
        # For now, we'll define the objects in the task
        pass

    def create_task(self):
        """Create the MTC task with all stages."""
        self.task = core.Task("pick_place_task")

        # Set task properties
        self.task.loadRobotModel(self.get_logger())

        # Get the robot model
        robot_model = self.task.getRobotModel()

        # Set task properties
        self.task.setProperty("group", self.arm_group_name)
        self.task.setProperty("eef", self.eef_name)
        self.task.setProperty("ik_frame", self.hand_frame)

        # ==========================
        # Create planning pipelines
        # ==========================

        # Sampling planner for complex motions
        sampling_planner = core.PipelinePlanner()
        sampling_planner.setProperty("goal_joint_tolerance", 1e-5)

        # Cartesian planner for linear motions
        cartesian_planner = core.CartesianPath()
        cartesian_planner.setMaxVelocityScalingFactor(1.0)
        cartesian_planner.setMaxAccelerationScalingFactor(1.0)
        cartesian_planner.setStepSize(0.01)

        # Interpolation planner for simple joint movements
        interpolation_planner = core.JointInterpolation()

        # ==========================
        # Stage 1: Current State
        # ==========================
        current_state = stages.CurrentState("current state")
        self.task.add(current_state)

        # ==========================
        # Stage 2: Open Hand (when gripper available)
        # ==========================
        # TODO: Enable when gripper/hand is integrated
        # open_hand = stages.MoveTo("open hand", interpolation_planner)
        # open_hand.setGroup(self.hand_group_name)
        # open_hand.setGoal("open")
        # self.task.add(open_hand)

        # ==========================
        # Stage 3: Move to Pick Position
        # ==========================
        pick_container = core.SerialContainer("pick object")
        self.task.add(pick_container)

        # 3a: Approach object
        approach = stages.MoveRelative("approach object", cartesian_planner)
        approach.setGroup(self.arm_group_name)

        # Set direction (move down in Z)
        approach.setMinMaxDistance(0.05, 0.15)

        # Define approach direction as Vector3Stamped
        direction = Vector3Stamped()
        direction.header.frame_id = "base_link"
        direction.vector.z = -1.0  # Move down
        approach.setDirection(direction)

        pick_container.insert(approach)

        # 3b: Generate grasp pose
        grasp_generator = stages.GenerateGraspPose("generate grasp pose")
        grasp_generator.setAngleDelta(0.2)  # Angular resolution for sampling
        grasp_generator.setPreGraspPose("open")  # Hand posture before grasp
        grasp_generator.setGraspPose("closed")   # Hand posture during grasp
        grasp_generator.setMonitoredStage(current_state)

        # Define object pose
        object_pose = PoseStamped()
        object_pose.header.frame_id = "base_link"
        object_pose.pose.position.x = 0.3
        object_pose.pose.position.y = 0.0
        object_pose.pose.position.z = 0.2
        object_pose.pose.orientation.w = 1.0

        grasp_generator.setObject(self.object_name)
        grasp_generator.setPose(object_pose)

        # Compute IK for grasp
        grasp_ik = stages.ComputeIK("grasp pose IK", grasp_generator)
        grasp_ik.setMaxIKSolutions(8)
        grasp_ik.setIKFrame(self.hand_frame)
        grasp_ik.setTargetPose(object_pose)
        grasp_ik.properties().configureInitFrom(core.Stage.PropertyInitializerSource.PARENT)

        pick_container.insert(grasp_ik)

        # 3c: Allow collision with object (approach stage)
        allow_touch = stages.ModifyPlanningScene("allow collision (hand-object)")
        allow_touch.allowCollisions(
            self.object_name,
            [self.hand_frame],  # Links that can touch the object
            True
        )
        pick_container.insert(allow_touch)

        # 3d: Close hand (when gripper available)
        # TODO: Enable when gripper/hand is integrated
        # close_hand = stages.MoveTo("close hand", interpolation_planner)
        # close_hand.setGroup(self.hand_group_name)
        # close_hand.setGoal("closed")
        # pick_container.insert(close_hand)

        # 3e: Attach object
        attach_object = stages.ModifyPlanningScene("attach object")
        attach_object.attachObject(self.object_name, self.hand_frame)
        pick_container.insert(attach_object)

        # 3f: Lift object
        lift = stages.MoveRelative("lift object", cartesian_planner)
        lift.setGroup(self.arm_group_name)
        lift.setMinMaxDistance(0.08, 0.15)

        # Lift direction (up in Z)
        lift_direction = Vector3Stamped()
        lift_direction.header.frame_id = "base_link"
        lift_direction.vector.z = 1.0  # Move up
        lift.setDirection(lift_direction)

        pick_container.insert(lift)

        # ==========================
        # Stage 4: Move to Place Position
        # ==========================

        # 4a: Move to pre-place pose
        move_to_place = stages.Connect(
            "move to place",
            [(self.arm_group_name, sampling_planner)]
        )
        move_to_place.properties().configureInitFrom(core.Stage.PropertyInitializerSource.PARENT)
        self.task.add(move_to_place)

        # ==========================
        # Stage 5: Place Object
        # ==========================
        place_container = core.SerialContainer("place object")
        self.task.add(place_container)

        # 5a: Lower object
        lower = stages.MoveRelative("lower object", cartesian_planner)
        lower.setGroup(self.arm_group_name)
        lower.setMinMaxDistance(0.05, 0.15)

        # Lower direction (down in Z)
        lower_direction = Vector3Stamped()
        lower_direction.header.frame_id = "base_link"
        lower_direction.vector.z = -1.0  # Move down
        lower.setDirection(lower_direction)

        place_container.insert(lower)

        # 5b: Generate place pose
        place_generator = stages.GeneratePlacePose("generate place pose")
        place_generator.setObject(self.object_name)

        # Define place location
        place_pose = PoseStamped()
        place_pose.header.frame_id = "base_link"
        place_pose.pose.position.x = 0.3
        place_pose.pose.position.y = 0.3
        place_pose.pose.position.z = 0.2
        place_pose.pose.orientation.w = 1.0

        place_generator.setPose(place_pose)
        place_generator.setMonitoredStage(current_state)

        place_container.insert(place_generator)

        # 5c: Open hand (when gripper available)
        # TODO: Enable when gripper/hand is integrated
        # open_hand_place = stages.MoveTo("open hand", interpolation_planner)
        # open_hand_place.setGroup(self.hand_group_name)
        # open_hand_place.setGoal("open")
        # place_container.insert(open_hand_place)

        # 5d: Detach object
        detach_object = stages.ModifyPlanningScene("detach object")
        detach_object.detachObject(self.object_name, self.hand_frame)
        place_container.insert(detach_object)

        # 5e: Forbid collision with object
        forbid_touch = stages.ModifyPlanningScene("forbid collision (hand-object)")
        forbid_touch.allowCollisions(
            self.object_name,
            [self.hand_frame],
            False
        )
        place_container.insert(forbid_touch)

        # 5f: Retreat
        retreat = stages.MoveRelative("retreat", cartesian_planner)
        retreat.setGroup(self.arm_group_name)
        retreat.setMinMaxDistance(0.05, 0.15)

        # Retreat direction (up in Z)
        retreat_direction = Vector3Stamped()
        retreat_direction.header.frame_id = "base_link"
        retreat_direction.vector.z = 1.0  # Move up
        retreat.setDirection(retreat_direction)

        place_container.insert(retreat)

        # ==========================
        # Stage 6: Return Home
        # ==========================
        return_home = stages.MoveTo("return home", sampling_planner)
        return_home.setGroup(self.arm_group_name)
        return_home.setGoal("home")  # Uses named pose from SRDF
        self.task.add(return_home)

        self.get_logger().info("MTC task created with all stages")

    def plan_task(self):
        """Plan the MTC task."""
        if self.task is None:
            self.get_logger().error("Task not created. Call create_task() first.")
            return False

        self.get_logger().info("Planning MTC task...")

        try:
            # Plan the task
            self.task.plan()

            # Check if planning succeeded
            if self.task.numSolutions() == 0:
                self.get_logger().error("No solutions found!")
                return False

            self.get_logger().info(f"Found {self.task.numSolutions()} solution(s)")
            return True

        except Exception as e:
            self.get_logger().error(f"Planning failed: {str(e)}")
            return False

    def execute_task(self):
        """Execute the planned MTC task."""
        if self.task is None:
            self.get_logger().error("Task not created.")
            return False

        if self.task.numSolutions() == 0:
            self.get_logger().error("No solutions to execute. Plan first.")
            return False

        self.get_logger().info("Executing MTC task...")

        try:
            # Execute the best solution
            self.task.execute(self.task.solutions()[0])
            self.get_logger().info("Task execution completed!")
            return True

        except Exception as e:
            self.get_logger().error(f"Execution failed: {str(e)}")
            return False

    def introspect_task(self):
        """Print task structure for debugging."""
        if self.task is None:
            self.get_logger().error("Task not created.")
            return

        self.get_logger().info("Task introspection:")
        self.get_logger().info(self.task.toString())


def main(args=None):
    """Main function to run the MTC demo."""
    rclpy.init(args=args)

    demo = MTCPickPlaceDemo()

    try:
        # Setup planning scene (add objects)
        demo.setup_planning_scene()

        # Create the task
        demo.create_task()

        # Print task structure
        demo.introspect_task()

        # Plan the task
        if demo.plan_task():
            # Ask user if they want to execute
            response = input("\nExecute the planned task? (y/n): ")
            if response.lower() == 'y':
                demo.execute_task()

        demo.get_logger().info("Demo completed")

    except KeyboardInterrupt:
        demo.get_logger().info("Demo interrupted by user")
    except Exception as e:
        demo.get_logger().error(f"Demo failed: {str(e)}")
    finally:
        demo.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
